Interfacing PODIO with Julia Final Blog GSOC 22 (#1232)
HSF/hsf.github.io: _gsocblogs/2022/blog_PODIO_SoumilBaldota.md

The prototype can be found at the below link
[Prototype](https://github.com/soumilbaldota/PODIO_Julia_Interface_Prototype.git)

After the prototype was created. I added an attribute to store the Julia type of a data member described in C++ to the MemberVariable Class and a function to [get the Julia type](https://github.com/AIDASoft/podio/pull/310/files#diff-c129698a9b29360c0e27c5e4f710981b4f99524ad44c039202d750bcf349c834) in [generator_utils.py](https://github.com/AIDASoft/podio/blob/julia/python/generator_utils.py), this was a necessary step to feed the jinja2 templates with the appropriate Julia type for code generation.
After Julia type was added I made some [unittests](https://github.com/AIDASoft/podio/pull/310/files#diff-61702c3a214182795b1d726c5dc1679a64a10274b7929e4ad4ceaf8dc87c203d) for the same and added them to [test_MemberParser.py](https://github.com/AIDASoft/podio/blob/julia/python/test_MemberParser.py).

> Following the addition of the attributes necessary for building the includes required to feed new templates that will generate Julia code. There is actual preprocessing required for building these includes required by our [Constructor](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/templates/Constructor.jl.jinja2) and the [MutableStruct](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/templates/MutableStruct.jl.jinja2) templates(covered further in the blog). This preprocessing is done by the [**preprocess_for_julia**](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/podio_class_generator.py#L295-L320) function. This functions fills our includes dictionary passed to the templates.

Reviewer: ```suggestion
Following the addition of the attributes necessary for building the includes required to feed new templates that will generate Julia code. There is actual preprocessing required for building these includes required by our [Constructor](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/templates/Constructor.jl.jinja2) and the [MutableStruct](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/templates/MutableStruct.jl.jinja2) templates (covered further in the blog). This preprocessing is done by the [**preprocess_for_julia**](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/podio_class_generator.py#L295-L320) function. This functions fills our includes dictionary passed to the templates.

```

> Mutually Recursive Type Declarations which are solved easily using forward declarations in C++ required the use of either Parametric types or Abstract Types in Julia([long standing issue](https://github.com/JuliaLang/julia/issues/269)), For the purpose of code generation and cleaner approach we chose to use the parameteric approach. This required the list of parameters required by each Constructor. This data is processed by the [**get_julia_params**](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/podio_class_generator.py#L286-L293)

Reviewer: ```suggestion
Mutually Recursive Type Declarations which are solved easily using forward declarations in C++ required the use of either Parametric types or Abstract Types in Julia ([long standing issue](https://github.com/JuliaLang/julia/issues/269)), For the purpose of code generation and cleaner approach we chose to use the parameteric approach. This required the list of parameters required by each Constructor. This data is processed by the [**get_julia_params**](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/podio_class_generator.py#L286-L293)

```

I think here you should add one sentence about the MutableStructs (I.e. the parametric types) and the constructors to 1) make the connection to the paragraph above and 2) to highlight the fact that we really try to hide this fact from the users by having the constructors with the expected name. It will then be much clearer to the reader what is actually happening and how the gemerated code looks like.

The Parameteric Types are used by the [Datatypes](https://github.com/AIDASoft/podio/blob/a8efda986f776892e90e7aafa4958aaf18a1e6c7/doc/templates.md?plain=1#L84), which have one to one and one to many relations and are not used by the [Components](https://github.com/AIDASoft/podio/blob/a8efda986f776892e90e7aafa4958aaf18a1e6c7/doc/templates.md?plain=1#L75) as they do not have any Cyclic Relations that could cause an infinite inclusion loop, Similarly Types with pure VectorMemberRelations also do not need a Parameteric Type.

The Datatypes described by the yaml input are allowed to have a namespaces. These namespaces have been implemented with the help of modules and Submodules in Julia. So for to emulate the namespaces we have grouped the code into modules, such that each Type is wrapped in its own module. For Example the MCParticle Constructor would be a part of the MCParticleModule. the modules generated in parent namespaces required the collection of children in each namespace this is handled by the function [**get_namespace_dict**](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/podio_class_generator.py#L107-L115). 

**Usage Of EDM4hep**

> 

Reviewer: Could you use the EDM4hep include here and then do e.g. edm4hep.MCParticle where necessary? That will probably be the most interesting use case for most people and you can still put the specific MCParticle related includes/imports at the bottom or into a separate code block.

project: HSF
title: Interfacing PODIO with Julia
author: Soumil Baldota
date: 4.09.2022 
year: 2022 
layout: blog_post
logo: podio-logo.png

> intro: Final progress report

Reviewer 2: The intro is what is visible on the front page with many posts, it is an incentive for visitors to read more, so it should not look like a title.

Reviewer 2: If you don't want to use a photo, we encourage using an avatar image of your choice, you can even use a web-based one such as your GitHub avatar that you can update later, see this [example](https://hepsoftwarefoundation.org/gsoc/blogs/2022/blog_HSF_TheAdmins.html)

Author: I have added recent commit addressing this.

project: HSF
title: Interfacing PODIO with Julia
author: Soumil Baldota
avatar: https://avatars.githubusercontent.com/u/71919921?s=400&v=4
date: 4.09.2022 
year: 2022 
layout: blog_post

> logo: podio-logo.png

Author: Added this to the latest commit

After Julia type was added I made some [unittests](https://github.com/AIDASoft/podio/pull/310/files#diff-61702c3a214182795b1d726c5dc1679a64a10274b7929e4ad4ceaf8dc87c203d) for the same and added them to [test_MemberParser.py](https://github.com/AIDASoft/podio/blob/julia/python/test_MemberParser.py).

Following the addition of the attributes necessary for building the includes required to feed new templates that will generate Julia code. There is actual preprocessing required for building these includes required by our [Constructor](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/templates/Constructor.jl.jinja2) and the [MutableStruct](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/templates/MutableStruct.jl.jinja2) templates(covered further in the blog). This preprocessing is done by the [**preprocess_for_julia**](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/podio_class_generator.py#L295-L320) function. This functions fills our includes dictionary passed to the templates.

> Mutually Recursive Type Declarations in Julia required the use of Parametric types which required the list of parameters required by each Constructor. This data is processed by the [**get_julia_params**](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/podio_class_generator.py#L286-L293)

Reviewer: Here you could also add a sentence that this is solved with forward declarations in c++ which is not possible in Julia. You can also link to the Julia issue describing the possible workarounds and maybe even briefly say why we chose the parametric types at the end.

You could also go into a few more details to describe which parts are now actually solved with parametric types (and why e.g. vector members are not).

Author: Added it in the latest commit

> Wrapping the modules generated in parent namespaces required the collection of children in each namespace this is handled by the function [**get_namespace_dict**](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/podio_class_generator.py#L107-L115). 

Reviewer: Maybe also add a brief sentence here why we want the grouping into modules in the first place.

Author: Added it in the latest commit

- [Constructor.jl.jinja2 template](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/templates/Constructor.jl.jinja2) is used for creating corresponding Constructors which initialize the structs with default values and bring the datatypes into their own namespace/module.

- [JuliaCollection.jl.jinja2 template](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/templates/JuliaCollection.jl.jinja2) is used for creating Collection of the given Datatype.

- [ParentModule.jl.jinja2 template](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/python/templates/ParentModule.jl.jinja2) is used to create a parent namespace.

**Testing**

[**unittests**](https://github.com/AIDASoft/podio/blob/59fe1e740ca0a7dbff1180c1425047ed1ab3e027/tests/unittest.jl) were added to test the functionality of the following in the generated code:
- Namespaces
- Collections
- Cyclic Dependency
- Vector Members
- Relations

> This was added to CI of the repo by Thomas.

Reviewer: ```suggestion
The unit test suite covering the Julia code generation of the example datamodel was added to the test setup with the help of the mentors. This setup is also run in the CI workflows if a suitable Julia version is detected.
```
